### ISSUE TITLE
Issue when number of simulations per data-generating mechanism are not of equal size

### ISSUE THREAD

**Issue by edbonneville** [ellessenne/rsimsum#30]
**Describe the bug**

Hi - thanks for the excellent package! I think there is an issue when running `simsum()` / `multisimsum()` on a dataset from a simulation study when  n<sub>sim</sub> differs per data-generating mechanism.

**To Reproduce**

Here is an example with the `frailty` dataset from the package:

```r
# Load library and data
library(rsimsum)
data("frailty", package = "rsimsum")

# Duplicate last row and and to data
frailty <- rbind.data.frame(frailty, frailty[nrow(frailty), ])

# Check nsim
with(frailty, table(model, interaction(fv_dist, par)))

model               Gamma.fv Log-Normal.fv Gamma.trt Log-Normal.trt
  Cox, Gamma            1000          1000      1000           1000
  Cox, Log-Normal       1000          1000      1000           1000
  RP(P), Gamma          1000          1000      1000           1000
  RP(P), Log-Normal     1000          1000      1000           1001
```

Attempt with `multisimsum()` :

```r
ms1 <- multisimsum(
  data = frailty,
  par = "par", 
  true = c(trt = -0.50, fv = 0.75),
  estvarname = "b",
  se = "se",
  ref = "Cox, Gamma",
  methodvar = "model",
  by = "fv_dist"
)

Error in stats::cor(data[[i]][[ref]][[estvarname]], data[[i]][[x]][[estvarname]],  : 
  incompatible dimensions
```

So it is getting stuck [here](https://github.com/ellessenne/rsimsum/blob/088780e77a5457f8acddbfad2fc5b382f4574aac/R/simsum.R#L179) it seems.. 

**Expected behaviour**

Should return a `multisimsum`  object with correct n<sub>sim</sub> values.

**System information:**

```r
R version 4.0.3 (2020-10-10)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 17134)

Matrix products: default

locale:
[1] LC_COLLATE=Dutch_Netherlands.1252  LC_CTYPE=Dutch_Netherlands.1252   
[3] LC_MONETARY=Dutch_Netherlands.1252 LC_NUMERIC=C                      
[5] LC_TIME=Dutch_Netherlands.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] rsimsum_0.9.1

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.5       dplyr_1.0.2      crayon_1.3.4     plyr_1.8.6       grid_4.0.3      
 [6] R6_2.5.0         ggridges_0.5.2   lifecycle_0.2.0  gtable_0.3.0     backports_1.2.0 
[11] magrittr_2.0.1   scales_1.1.1     pillar_1.4.7     ggplot2_3.3.2    rlang_0.4.9     
[16] rstudioapi_0.13  generics_0.1.0   vctrs_0.3.6      checkmate_2.0.0  ellipsis_0.3.1  
[21] tools_4.0.3      glue_1.4.2       purrr_0.3.4      munsell_0.5.0    compiler_4.0.3  
[26] pkgconfig_2.0.3  colorspace_2.0-0 tidyselect_1.1.0 tibble_3.0.4
```

**Additional context**

I think it would be a welcome fix since the number of simulated datasets will likely vary if for example you vary the sample size per scenario, in which case lower values of sample size would need larger n<sub>sim</sub> to achieve the same Monte-Carlo error for some parameter :-) 

Thanks!
-Ed

**Comment by ellessenne** [comment]
Hi @edbonneville, and thanks for your feedback!

Does that scenario actually happen in practice? I mean, that corresponds to a DGM where you fit only one of the methods? 
In principle, I would assume that (per replication of a DGM) you get a _block_ of data with estimates from every model, something like:

``` r
# Load library and data
library(rsimsum)
data("frailty", package = "rsimsum")

# Duplicate last row and and to data
frailty <- rbind.data.frame(frailty, frailty[(nrow(frailty)-3):nrow(frailty), ])

# Check nsim
with(frailty, table(model, interaction(fv_dist, par)))
#>                    
#> model               Gamma.fv Log-Normal.fv Gamma.trt Log-Normal.trt
#>   Cox, Gamma            1000          1000      1000           1001
#>   Cox, Log-Normal       1000          1000      1000           1001
#>   RP(P), Gamma          1000          1000      1000           1001
#>   RP(P), Log-Normal     1000          1000      1000           1001
```

...which works:

```r
ms1 <- multisimsum(
    data = frailty,
    par = "par",
    true = c(trt = -0.50, fv = 0.75),
    estvarname = "b",
    se = "se",
    ref = "Cox, Gamma",
    methodvar = "model",
    by = "fv_dist"
)
ms1
#> 
#> Estimands variable: par 
#>  Unique estimands: fv, trt 
#>  True values: fv = -0.5, trt = 0.75 
#> 
#> Method variable: model 
#>  Unique methods: Cox, Gamma, Cox, Log-Normal, RP(P), Gamma, RP(P), Log-Normal 
#>  Reference method: Cox, Gamma 
#> 
#> By factors: fv_dist 
#> 
#> Monte Carlo standard errors were computed.
```

Am I misunderstanding the scenario you are suggesting adding?

**Comment by edbonneville** [comment]
Thanks @ellessenne , you are right that the block of data per DGM is what you would normally expect.

In the meantime, I did actually find what went wrong when trying to use `simsum()` on the dataset I am using (forgive my earlier reprex!): I had missed that one of the "by" variables I was using was actually *nested* in the "method" variable. In my case, I was comparing multiple imputation (while varying the number of imputations) with complete case analysis - and I used the number of imputations as a "by" variable.

Better illustrated here:

``` r
# Load library and set n_sim
library(rsimsum)
n_sim <- 100

# Compare complete case analysis (CCA), and mice (with 5 or 10 imputations)
n_imps <- c(5, 10)

dat <- cbind.data.frame(
  "method" = c(rep("CCA", n_sim), rep("mice", length(n_imps) * n_sim)),
  "n_imp" = c(rep(0, n_sim), rep(n_imps[1], n_sim), rep(n_imps[2], n_sim))
)

# Add fake estimates/se's
dat$estimate <- rnorm(length(dat))
dat$se <- runif(length(dat), min = 0.05, max = 0.10)

head(dat)
#>   method n_imp   estimate         se
#> 1    CCA     0 -1.8268701 0.05122508
#> 2    CCA     0  0.2234982 0.09482914
#> 3    CCA     0 -1.8268701 0.08986567
#> 4    CCA     0  0.2234982 0.05122508
#> 5    CCA     0 -1.8268701 0.09482914
#> 6    CCA     0  0.2234982 0.08986567

# Try to use "n_imp" as by var - will fail
s1 <- simsum(
  data = dat,
  estvarname = "estimate",
  se = "se",
  true = 0,
  methodvar = "method",
  by = "n_imp"
)
#> 'ref' method was not specified, CCA set as the reference
#> Error in stats::cor(data[[i]][[ref]][[estvarname]], data[[i]][[x]][[estvarname]], : incompatible dimensions

# Make better "method" variable
dat$new_method <- paste(dat$method, as.character(dat$n_imp), sep = "_")

# Will now work
s2 <- simsum(
  data = dat,
  estvarname = "estimate",
  se = "se",
  true = 0,
  methodvar = "new_method"
)
#> 'ref' method was not specified, CCA_0 set as the reference
```

While this does work, in this case it feels more natural (to me) to have the number of imputations as a "by" variable - particularly if I was planning to vary this with many different values , where the method variable would then have a lot of levels. I guess the simple solution to this (using the reprex example) would be to run it like this and then extract `s2$summ` and separate the "new_method" column out into columns "method" and "n_imp" ..

**Comment by ellessenne** [comment]
Aha, I understand now - yeah I think the issue there is that the nesting throws the _splitting_ of the data by `methodvar`-`by` off. No worries about the previous reprex, that does in fact break so it's good to know! 😃

I actually think that in your example the number of multiple imputations is something intrinsically related to the method and not to the data-generating mechanism itself, which focusses on generating data according to a given model (e.g. MAR, MNAR, etc. in the missing data example). 
...so I would actually use the second approach you coded above to set-up the study with a larger number of methods! 😅 
And yeah, then maybe split the raw results data (p.s.: there's a helper function for to extract the data, try e.g. `get_data(s2)`) to some some custom-made plot or table. Does it make any sense?

I am however wondering if there should be a _soft_ fail (maybe with a warning), as that calculation is only used to obtain the relative precision of a method vs another (if I remember correctly, which might not be always of interest). What do you think? Or do you have any other example of simulation study where there might be some uneven nesting like that?

**Comment by edbonneville** [comment]
Yes would agree that the number of imputations are not related to the DGMs; I just hadn't thought about this particular nesting since I was computing the performance measures on my own with data.table and had just put the number of imputations in `by = ...` without giving it a second thought!

I think for the soft fail it is up to you; on one hand it would be nice for the procedure not to fail due to just the calculation of relative precision, on the other I cannot say that I have other examples of uneven nesting off the top of my head.. I don't think it is an urgent fix in any case 😃

**Comment by ellessenne** [comment]
Hi @edbonneville, sorry for not getting back to you earlier!
I was thinking a bit about this, and realised that this would be solved by implementing #24 (which has been in the back of my mind for months now). 
Do you think that would work?

**Comment by edbonneville** [comment]
No worries @ellessenne - I think it is a good idea! There are definitely more examples of method related properties that one might choose to vary and that may/may not lead to uneven nesting; off the top of my head:

- If you maybe want to compare a standard Cox model with a piecewise-constant hazards model, you could choose to vary the number of knots in the latter (= uneven nesting)
- You might be validating a bunch of models in a sim study and you could choose to vary the number of cross-validation folds if that was relevant (= even nesting)

**Comment by ellessenne** [comment]
Yep, that makes sense - however, the nesting between methods _factors_ would not matter there, as the underlying code would compute a Cartesian product of all possible methods-factors combinations and use that as a _method_ identifier. And of course only combinations that appear in `data` would be kept. 
That makes sense, right?

**Comment by edbonneville** [comment]
Sounds good to me :-)

**Comment by ellessenne** [comment]
I'm closing this, in principle a solution is implemented in #31!
